src/cascadia/TerminalControl/XamlLights.cpp

```diff
@@ -0,0 +1,107 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+#include "pch.h"
+#include "TermControl.h"
+#include "XamlLights.h"
+#include "VisualBellLight.g.cpp"
+
+using namespace winrt::Windows::UI::Xaml;
+using namespace winrt::Windows::UI::Xaml::Media;
+
+namespace winrt::Microsoft::Terminal::Control::implementation
+{
+    DependencyProperty VisualBellLight::_IsTargetProperty{ nullptr };
+
+    VisualBellLight::VisualBellLight()
+    {
+        _InitializeProperties();
+    }
+
+    void VisualBellLight::_InitializeProperties()
+    {
+        // Initialize any dependency properties here.
+        // This performs a lazy load on these properties, instead of
+        // initializing them when the DLL loads.
+        if (!_IsTargetProperty)
+        {
+            _IsTargetProperty =
+                DependencyProperty::RegisterAttached(
+                    L"IsTarget",
+                    winrt::xaml_typename<bool>(),
+                    winrt::xaml_typename<Control::VisualBellLight>(),
+                    PropertyMetadata{ winrt::box_value(false), PropertyChangedCallback{ &VisualBellLight::OnIsTargetChanged } });
+        }
+    }
+
+    // Method Description:
+    // - This function is called when the first target UIElement is shown on the screen,
+    //   this enables delaying composition object creation until it's actually necessary.
+    // Arguments:
+    // - newElement: unused
+    void VisualBellLight::OnConnected(UIElement const& /* newElement */)
+    {
+        if (!CompositionLight())
+        {
+            auto spotLight{ Window::Current().Compositor().CreateAmbientLight() };
+            spotLight.Color(Windows::UI::Colors::White());
+            CompositionLight(spotLight);
+        }
+    }
+
+    // Method Description:
+    // - This function is called when there are no more target UIElements on the screen
+    // - Disposes of composition resources when no longer in use
+    // Arguments:
+    // - oldElement: unused
+    void VisualBellLight::OnDisconnected(UIElement const& /* oldElement */)
+    {
+        if (CompositionLight())
+        {
+            CompositionLight(nullptr);
+        }
+    }
+
+    winrt::hstring VisualBellLight::GetId()
+    {
+        return VisualBellLight::GetIdStatic();
+    }
+
+    void VisualBellLight::OnIsTargetChanged(DependencyObject const& d, DependencyPropertyChangedEventArgs const& e)

```

Comment:
I don't think I totally understand what's going on with this method. This is a static function that's called when `IsTarget` changes on a `VisualBellLight`? And then `d` is either a `UIElement`, or a `Brush`... When is this ever a brush?

Or is this just forward thinking? All this seems like handy boilerplate around `XamlLight`, I'm kinda shocked (but also not surprised) that we need this much boilerplate just to add a light.

Reply:
So yeah I think this is some cool stuff -- he's created an "attached property" (like `Grid.Row` -- a property that's about one object, stored in global storage, attached to another object). It looks like this one can be attached to either a brush or a UI Element, seamlessly.

Reply:
Yep! I've mostly just adopted the code from the [xaml light docs](https://docs.microsoft.com/en-us/windows/uwp/composition/xaml-lighting). Lights can only target a `Brush` or a `UIElement`, and yeah in our case we never use it for a brush but I feel like its probably a good idea to have that check anyway